internal/handlers/common/error.go

```diff
@@ -108,21 +139,87 @@ func (e *Error) Document() *types.Document {
 	return d
 }
 
-// ProtocolError converts any error to wire protocol error.
-//
-// Nil panics, *Error (possibly wrapped) is returned unwrapped with true,
-// any other value is wrapped with InternalError and returned with false.
-func ProtocolError(err error) (*Error, bool) {
-	if err == nil {
-		panic("err is nil")
+// WriteErrors represents slice of protocol write errors.
+type WriteErrors []WriteError
+
+// NewWriteErrorMsg creates new protocol write error with given ErrorCode and message.
+func NewWriteErrorMsg(code ErrorCode, msg string) error {
+	return &WriteErrors{{
+		code: code,
+		err:  errors.New(msg),
+	}}
+}
+
+// Error implements error interface.
+func (we *WriteErrors) Error() string {
+	var err string
+	for _, e := range *we {
+		err += e.Error() + ","
 	}
 
-	var e *Error
-	if errors.As(err, &e) {
-		return e, true
+	return err
+}
+
+// Code implements ProtoErr interface.
+func (we *WriteErrors) Code() ErrorCode {
+	for _, e := range *we {
+		return e.code
 	}
+	return errUnset
+}
 
-	return NewError(errInternalError, err).(*Error), false
+// Unwrap implements standard error unwrapping interface.
+func (we *WriteErrors) Unwrap() error {
+	for _, e := range *we {
+		return &e
+	}
+	return nil
+}
+
+// Document implements ProtoErr interface..
+func (we *WriteErrors) Document() *types.Document {
+	errs := must.NotFail(types.NewArray())
+	for _, e := range *we {
+		must.NoError(errs.Append(e.Document()))
+	}
+
+	d := must.NotFail(types.NewDocument(
+		"ok", float64(1),
+		"writeErrors", errs,
+	))
+	return d
+}
+
+// WriteError represents protocol write error.
+// It required to build the correct write error result.
+type WriteError struct {
+	code ErrorCode
+	err  error
+}
+
+// Error implements error interface.
+func (we *WriteError) Error() string {
+	return fmt.Sprintf("%[1]s (%[1]d): %[2]v", we.code, we.err)
+}
+
+// Code implements ProtoErr interface.
+func (we *WriteError) Code() ErrorCode {
+	return we.code
+}
+
+// Unwrap implements standard error unwrapping interface.
+func (we *WriteError) Unwrap() error {
+	return we.err
+}
+
+// Document implements ProtoErr interface.
+// Fields "code" and "errmsg" must always be filled in so that clients can parse the error message.
+func (we *WriteError) Document() *types.Document {
```

Comment:
Could you please add a comment that otherwise it would parse it as a command error?

Reply:
Added.

Reply:
Thank you.

> Added.